### PR TITLE
K8SPSMDB-325: Do not reconcile cluster if Paused

### PR DESCRIPTION
**Pull request by blez**
[K8SPSMDB-325](https://jira.percona.com/browse/K8SPSMDB-325)

### CODE REVIEW

File: pkg/controller/perconaservermongodb/mgo.go [hunk 1]
```diff
@@ -38,6 +38,11 @@ var errReplsetLimit = fmt.Errorf("maximum replset member (%d) count reached", mo
 
 func (r *ReconcilePerconaServerMongoDB) reconcileCluster(cr *api.PerconaServerMongoDB, replset *api.ReplsetSpec,
 	pods corev1.PodList, usersSecret *corev1.Secret, mongosPods []corev1.Pod) (mongoClusterState, error) {
+
+	if cr.Spec.Pause {
```
Comment (delgod) [inline pkg/controller/perconaservermongodb/mgo.go:42]: I think situation `cr.Status.Replsets[replset.Name].Size: 0` and `cr.Spec.Pause: flase` is possible.
I would prefer to check `cr.Status.Replsets[replset.Name].Size: 0` check

**Review by delgod — CHANGES_REQUESTED** [review]

File: pkg/controller/perconaservermongodb/mgo.go (hunk 1 above)
Comment (blez) [reply pkg/controller/perconaservermongodb/mgo.go]: https://github.com/percona/percona-server-mongodb-operator/pull/477/commits/2e16879713cda7825b44d1755fdf338f2576c794

**Review by tplavcic — CHANGES_REQUESTED** [review]
This looks ok now that it doesn't try to contact mongodb, but the cluster state is `ready` which probably shouldn't be if there is no cluster running:
```
  replsets:
    cfg:
      initialized: true
      ready: 0
      size: 0
      status: ready
    rs0:
      added_as_shard: true
      initialized: true
      ready: 0
      size: 0
      status: ready
  state: ready
```

**Comment by blez** [comment]
@tplavcic , @delgod Should we create new status for such case?
`Paused` maybe?

**Comment by tplavcic** [comment]
> @tplavcic , @delgod Should we create new status for such case?
> `Paused` maybe?

@blez That sounds reasonable to me, but to be honest I have no idea at the current moment if this can have some other implications in the cluster/operator lifecycle.

**Comment by delgod** [comment]
> @tplavcic , @delgod Should we create new status for such case?
> 
> `Paused` maybe?

up to @spron-in, Platform always expect status Ready

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| storage | passed |
| upgrade | passed |
| self-healing | passed |
| limits | passed |
| upgrade-consistency | passed |
| scaling | passed |
| demand-backup | passed |
| security-context | passed |
| operator-self-healing | passed |
| scheduled-backup | passed |
| demand-backup-sharded | passed |

**Comment by JNKPercona** [comment]
PSMDB operator docker - perconalab/percona-server-mongodb-operator:PR-477-d818b65f

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| storage | passed |
| one-pod | passed |
| upgrade | passed |
| self-healing | passed |
| monitoring | passed |
| limits | passed |
| upgrade-consistency | passed |
| scaling | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| security-context | passed |
| operator-self-healing | passed |
| scheduled-backup | passed |
| arbiter | passed |
| service-per-pod | passed |
| demand-backup-sharded | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |

**Comment by JNKPercona** [comment]
PSMDB operator docker - perconalab/percona-server-mongodb-operator:PR-477-6787204b

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| init-deploy | passed |
| one-pod | passed |
| upgrade | passed |
| self-healing | passed |
| monitoring | passed |
| limits | passed |
| upgrade-consistency | passed |
| scaling | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| security-context | passed |
| operator-self-healing | passed |
| scheduled-backup | passed |
| arbiter | passed |
| demand-backup-sharded | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |

**Comment by spron-in** [comment]
> > @tplavcic , @delgod Should we create new status for such case?
> > `Paused` maybe?
> 
> up to @spron-in, Platform always expect status Ready

Love the idea on new status, but not in this release.

**Comment by qJkee** [comment]
Can we make such check at the beginning of the `Reconcile` function?
It will reduce load to k8s and we will do less job, such as `CheckNSetDefaults`, `reconcileUsersSecret` and etc

File: pkg/controller/perconaservermongodb/smart.go [hunk 2]
```diff
@@ -24,6 +24,10 @@ import (
 func (r *ReconcilePerconaServerMongoDB) smartUpdate(cr *api.PerconaServerMongoDB, sfs *appsv1.StatefulSet,
 	replset *api.ReplsetSpec, secret *corev1.Secret) error {
 
+	if cr.Spec.Pause {
```
Comment (delgod) [inline pkg/controller/perconaservermongodb/smart.go:27]: I think the same here
```suggestion
	if replset.Size == 0 {
```

**Comment by blez** [comment]
@qJkee We could, but it will cause more conditions to check.
When user apply `pause: true` we would have to get current number of pods for each replica set and mongos. If there is no pods than we can stop at this point. If there are a few we should proceed to reconcile to be sure that operator set `size = 0` for all replicasets and mongos.
And it's just one example that came to my mind. Maybe there is more corner cases.
@delgod, what do you think?

File: pkg/controller/perconaservermongodb/smart.go [hunk 3]
```diff
@@ -196,7 +200,7 @@ func (r *ReconcilePerconaServerMongoDB) isAllSfsUpToDate(cr *api.PerconaServerMo
 }
 
 func (r *ReconcilePerconaServerMongoDB) startBalancerIfNeeded(cr *api.PerconaServerMongoDB, username, password string) error {
-	if !cr.Spec.Sharding.Enabled {
+	if !cr.Spec.Sharding.Enabled || cr.Spec.Pause {
```
Comment (delgod) [inline pkg/controller/perconaservermongodb/smart.go:203]:
```suggestion
	if !cr.Spec.Sharding.Enabled || cr.Spec.Sharding.Mongos.Size == 0 {
```

**Review by delgod — CHANGES_REQUESTED** [review]

**Comment by JNKPercona** [comment]
PSMDB operator docker - perconalab/percona-server-mongodb-operator:PR-477-615b1eb2

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| storage | passed |
| one-pod | passed |
| upgrade | passed |
| self-healing | passed |
| monitoring | passed |
| limits | passed |
| upgrade-consistency | passed |
| scaling | passed |
| demand-backup | passed |
| monitoring-2-0 | passed |
| operator-self-healing | passed |
| security-context | passed |
| arbiter | passed |
| scheduled-backup | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| demand-backup-sharded | passed |
| users | passed |

**Comment by JNKPercona** [comment]
PSMDB operator docker - perconalab/percona-server-mongodb-operator:PR-477-f6ce1a4e

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| one-pod | passed |
| init-deploy | passed |
| upgrade | passed |
| self-healing | passed |
| monitoring | passed |
| limits | passed |
| upgrade-consistency | passed |
| monitoring-2-0 | passed |
| scaling | passed |
| demand-backup | passed |
| operator-self-healing | passed |
| security-context | passed |
| arbiter | passed |
| service-per-pod | passed |
| scheduled-backup | passed |
| liveness | passed |
| demand-backup-sharded | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |

**Review by delgod — APPROVED** [review]

**Comment by JNKPercona** [comment]
PSMDB operator docker - perconalab/percona-server-mongodb-operator:PR-477-eeb83e1b

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| one-pod | passed |
| init-deploy | passed |
| upgrade | passed |
| monitoring | passed |
| limits | passed |
| scaling | passed |
| monitoring-2-0 | passed |
| self-healing | passed |
| security-context | passed |
| arbiter | passed |
| operator-self-healing | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| upgrade-consistency | passed |
| demand-backup | passed |
| users | passed |
| scheduled-backup | passed |
| demand-backup-sharded | passed |

**Review by tplavcic — APPROVED** [review]

**Review by nonemax — APPROVED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| init-deploy | passed |
| one-pod | passed |
| upgrade | passed |
| self-healing | passed |
| monitoring | passed |
| limits | passed |
| upgrade-consistency | passed |
| scaling | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| operator-self-healing | passed |
| security-context | passed |
| arbiter | passed |
| scheduled-backup | failed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| demand-backup-sharded | passed |
| users | passed |

**Comment by JNKPercona** [comment]
PSMDB operator docker - perconalab/percona-server-mongodb-operator:PR-477-6f97084b

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| self-healing | passed |
| operator-self-healing | passed |
| init-deploy | passed |
| limits | passed |
| scaling | passed |
| one-pod | passed |
| security-context | passed |
| monitoring | passed |
| monitoring-2-0 | passed |
| arbiter | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |
| upgrade | passed |
| upgrade-consistency | passed |
| demand-backup | passed |
| scheduled-backup | passed |
| demand-backup-sharded | passed |

**Pull request merged**